### ISSUE TITLE
During drag-and-drop, expanding a folder doesn't load its contents

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#990]
This is perhaps related to #977.  When a folder has already been loaded, and a new item is held over it for half a second, it opens up so that the folder can be placed inside.  This is expected.  However, this doesn't work unless the folder's contents were already loaded.  That means that to move a project into a nested folder, one must first manually expand all the parent folders, then drag it.

**Comment by nixorn** [comment]
In progress.

**Comment by Powell-v2** [comment]
Loading folder's contents doesn't seem to be feasible with the current setup. As far as I can tell, `jqTree` plugin doesn't offer any applicable events, and defining custom handlers doesn't work either.

How about loading all nested projects initially? This way, folders will be expanded on hover. As an added bonus, this opens doors to restoring most recent tree state on page reload.

**Comment by cdsmith** [comment]
Yes, I like the idea of loading the whole directory tree at once, then.  It's not likely to be large, and this will remove a lot of weird edge cases.

However, it will still be important to refresh the view occasionally, since it's possible that a user is logged in through two computers at once.  So we could still perhaps listen for the event when a folder is expanded, and reload the contents of that folder.

**Comment by cdsmith** [comment]
I don't recall if there's an endpoint on the server to load the whole tree at once.  If not, I can add one.

**Comment by Powell-v2** [comment]
I have no idea if one exists, but [we already touched on this before](https://github.com/google/codeworld/issues/1371#issuecomment-647482638).

**Comment by cdsmith** [comment]
I have now added '/listFolders' to the API, which should unblock client-side changes for this one.

**Comment by Powell-v2** [comment]
> we could still perhaps listen for the event when a folder is expanded, and reload the contents of that folder.

This could certainly be implemented, but existing response from the `listFolder` endpoint doesn't return nested items beyond first level. Would it be possible to adjust it so it hands back all nested content under a given folder?

**Comment by cdsmith** [comment]
This is already done in https://github.com/google/codeworld/commit/0c28ad4651185f5290c6df8f3cbd2cd9ebedfcc0

Not that the old endpoint, `/listFolder`, still returns only one level.  I didn't change it just to be sure existing code didn't break.  There is a new endpoint, `/listFolders` (note the s at the end) that includes all children.

**Comment by Powell-v2** [comment]
Oh I see, thought `listFolders` is only capable of returning entire directory tree at once. Thanks for clarifying!
